internal/util/k8sutil/crd.go

```diff
@@ -53,10 +58,96 @@ func GetCRDManifestPaths(crdsDir string) (crdPaths []string, err error) {
 		if info == nil {
 			return nil
 		}
-		if !info.IsDir() && strings.HasSuffix(path, "_crd.yaml") {
-			crdPaths = append(crdPaths, path)
+		if !info.IsDir() {
+			b, err := ioutil.ReadFile(path)
+			if err != nil {
+				return errors.Wrapf(err, "error reading manifest %s", path)
+			}
+			typeMeta, err := GetTypeMetaFromBytes(b)
+			if err != nil {
+				return errors.Wrapf(err, "error getting kind from manifest %s", path)
+			}
+			if typeMeta.Kind == "CustomResourceDefinition" {
+				crdPaths = append(crdPaths, path)
+			}
 		}
 		return nil
 	})
 	return crdPaths, err
 }
+
+// ParseGroupSubpackages parses the apisDir directory tree and returns a map of
+// all found API groups to subpackages.
+func ParseGroupSubpackages(apisDir string) (map[string][]string, error) {
+	return parseGroupSubdirs(apisDir, false)
+}
+
+// ParseGroupSubdirs parses the apisDir directory tree and returns a map of
+// all found API groups to versions.
+func ParseGroupVersions(apisDir string) (map[string][]string, error) {
+	return parseGroupSubdirs(apisDir, true)
+}
+
+var versionRegexp = regexp.MustCompile("^v[1-9][0-9]*((alpha|beta)[1-9][0-9]*)?$")
+
+func parseGroupSubdirs(apisDir string, strict bool) (map[string][]string, error) {
+	gvs := make(map[string][]string)
+	groups, err := ioutil.ReadDir(apisDir)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error reading directory %q to find API groups", apisDir)
+	}
+
+	for _, g := range groups {
+		if g.IsDir() {
+			groupDir := filepath.Join(apisDir, g.Name())
+			versions, err := ioutil.ReadDir(groupDir)
+			if err != nil {
+				return nil, errors.Wrapf(err, "error reading directory %q to find API versions", groupDir)
+			}
+
+			gvs[g.Name()] = make([]string, 0)
+			for _, v := range versions {
+				if v.IsDir() {
+					// Ignore directories that do not contain any files, so generators
+					// do not get empty directories as arguments.
+					verDir := filepath.Join(groupDir, v.Name())
+					files, err := ioutil.ReadDir(verDir)
+					if err != nil {
+						return nil, errors.Wrapf(err, "error reading directory %q to find API source files", verDir)
+					}
+					for _, f := range files {
+						if !f.IsDir() && filepath.Ext(f.Name()) == ".go" {
+							vsplit := strings.Split(v.Name(), string(filepath.Separator))
```

Comment:
Why do we need to split a potential version name `v1alpha1` on a separator `/`?
`v` isn't a filepath right?

Also any reason why we can't move this inside the condition `if strict {`? `vsplit` is only being used there.

Reply:
True, the version dir should not contain any separators since its from `ioutil.ReadDir()`, which does a shallow read. This bit can use `v.Name()` directly.